src/Parser.ts

```diff
@@ -807,7 +807,7 @@ export abstract class Parser extends Recognizer<Token, ParserATNSimulator> {
 			let ruleIndex: number = p.ruleIndex;
 			if (ruleIndex < 0) stack.push("n/a");
 			else stack.push(ruleNames[ruleIndex]);
-			p = p._parent;
+			p = p._parent as RuleContext;
```

Comment:
> Yes, I tried a local variable and that worked without problem too. I liked this version because it touched less lines of code.

The form used here is incorrect. You assert that `p._parent` is a `RuleContext` (not `undefined`), but if that is always the case then the code is an infinite loop.

> I'm trying to put together an isolated repro case to file a bug with the TypeScript team.

Here is a self-contained reproducer:

```ts
var obj: Context = {
  parent: undefined
};

function working(p: Context | undefined) {
  while (p != null) {
    p = p.parent;
  }
}

function broken(p: Context | undefined = obj) {
  while (p != null) {
    p = p.parent;
  }
}

interface Context { parent: Context | undefined; }
```